test/extended/util/framework.go

```diff
@@ -186,72 +185,92 @@ var CheckImageStreamTagNotFoundFn = func(i *imageapi.ImageStream) bool {
+	// the predicates isOK or isFailed, and the associated error in case of
+	// failure.
+	okOrFailed := func(rc *kapi.ReplicationController) (err error, matched bool) {
+		if isOK(rc) {
+			return nil, true
+		}
+		if isFailed(rc) {
+			return fmt.Errorf("The deployment %q status is %q", name, rc.Annotations[deployapi.DeploymentStatusAnnotation]), true
+		}
+		return nil, false
+	}
+
+	// waitForDeployment waits until okOrFailed returns true or the done
+	// channel is closed.
+	waitForDeployment := func() (err error, retry bool) {
 		requirement, err := labels.NewRequirement(deployapi.DeploymentConfigAnnotation, labels.EqualsOperator, sets.NewString(name))
 		if err != nil {
-			return fmt.Errorf("unexpected error generating label selector: %v", err)
+			return fmt.Errorf("unexpected error generating label selector: %v", err), false
 		}
-
 		list, err := client.List(kapi.ListOptions{LabelSelector: labels.NewSelector().Add(*requirement)})
 		if err != nil {
-			return err
+			return err, false
 		}
-
-		for i := range list.Items {
-			if isOK(&list.Items[i]) {
-				return nil
-			}
-			if isFailed(&list.Items[i]) {
-				return fmt.Errorf("The deployment %q status is %q",
-					name, list.Items[i].Annotations[deployapi.DeploymentStatusAnnotation])
+		for _, rc := range list.Items {
+			err, matched := okOrFailed(&rc)
+			if matched {
+				return err, false
 			}
 		}
-
-		rv := list.ResourceVersion
-		w, err := client.Watch(kapi.ListOptions{LabelSelector: labels.NewSelector().Add(*requirement), ResourceVersion: rv})
+		w, err := client.Watch(kapi.ListOptions{LabelSelector: labels.NewSelector().Add(*requirement), ResourceVersion: list.ResourceVersion})
 		if err != nil {
-			return err
+			return err, false
 		}
 		defer w.Stop()
-
 		for {
-			timeout := make(chan bool, 1)
-			go func() {
-				time.Sleep(10 * time.Second)
-				timeout <- true
-			}()
-
-			var val watch.Event
-			var ok bool
-
 			select {
-			case val, ok = <-w.ResultChan():
-			case <-timeout:
-			}
-
-			if !ok {
-				// reget and re-watch
-				break
-			}
-			if e, ok := val.Object.(*kapi.ReplicationController); ok {
-				if isOK(e) {
-					return nil
+			case val, ok := <-w.ResultChan():
+				if !ok {
+					// watcher error, re-get and re-watch
+					return nil, true
 				}
-				if isFailed(e) {
-					return fmt.Errorf("The deployment %q status is %q",
-						name, e.Annotations[deployapi.DeploymentStatusAnnotation])
+				if rc, ok := val.Object.(*kapi.ReplicationController); ok {
+					err, matched := okOrFailed(rc)
+					if matched {
+						return err, false
+					}
 				}
+			case <-done:
+				// no more time left, stop what we were doing,
+				// do no retry.
+				return nil, false
 			}
 		}
+	}
+
+	// errCh is buffered so the goroutine below never blocks on sending,
+	// preventing a goroutine leak if we reach the timeout.
+	errCh := make(chan error, 1)
+
+	go func() {
+		defer close(errCh)
+		err, retry := waitForDeployment()
```

Comment:
```go
defer close(errCh)
for {
    err, retry := waitForDeployment()
    if !retry {
        break
    }
}
errCh <- err
```

The above is wrong because `err` is not defined in the last line.